### ISSUE TITLE
Attributes sent without AttributeValues

### ISSUE THREAD

**Issue by gtkrug** [simplesamlphp/simplesamlphp#282]
Sometimes the attribute value for one of the attributes asserted by my Simple SAML IDP has no value in the user store (DB).  When this happens the SAML Assertion includes the saml:Attribute but the saml:AttributeValue is empty, for example:

    <saml:Attribute Name="sample" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
       <saml:AttributeValue xsi:type="xs:string"/>
     </saml:Attribute>

**Comment by thijskh** [comment]
We see this as well and (coincidentally?) also with the SQL module as the user store. Such values should indeed not be sent out because the SAML spec does not allow it. "Within an `<AttributeStatement>`, if the SAML attribute exists but has no values, then the `<AttributeValue>` element MUST be omitted."

**Comment by thijskh** [comment]
I did a bit more research. I could reproduce the above XML output in the SAML Tracer, but it appears to be an artefact of the XML pretty printing. What is actually sent in the assertion (when you decode the base64) is:
```xml
<saml:Attribute Name="sample" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
    <saml:AttributeValue xsi:type="xs:string"></saml:AttributeValue>
</saml:Attribute>
```
So if the database has an empty string in a given field, exactly that empty string gets sent. This seems the correct behaviour to me. When the field is in fact NULL, then the attribute is not sent for this user. This also makes sense to me.

Do you agree @gtkrug?

**Comment by thijskh** [comment]
I did create a unit test in saml2 to confirm and to guarantee for the future that creating an empty string AttributeValue results in `<saml:AttributeValue xsi:type="xs:string"></saml:AttributeValue>`.

**Comment by gtkrug** [comment]
I agree that there is a difference between an empty string and NULL, but personally I do not think an empty string should be sent, although I believe that depends on the semantics of the attribute.  An attribute could be defined that has no values, and simply the presence or absence of it is all that matters to an SP, but for the case where the semantics of the attribute do not allow for an empty string, I think there is value in having a filter that would remove any attribute without a value.

**Comment by gtkrug** [comment]
It might be that this is less a bug and more a feature request I suppose.

**Comment by thijskh** [comment]
With me that begs the question that if you have an attribute that cannot semantically be an empty string, why it is an empty string in the DB instead of NULL.
